### ISSUE TITLE
OpenVR Dynamic Resolution does not work with Unreal Engine 4 games

### ISSUE THREAD

**Issue by Crimson-foxGITHUB** [Louka3000/OpenVR-Dynamic-Resolution#16]
I decided to test Propagation VR with your application running, and it didn't change the resolution at all, even when increasing graphics settings to the maximum. UE4 handles resolution scaling stuffs a little differently, and I've noticed that not a single working title listed in WorkingGames.md as of typing this runs on Unreal Engine 4: every single game listed uses Unity in some form.

**Comment by Louka3000** [comment]
Was https://github.com/Louka3000/OpenVR-Dynamic-Resolution/issues/15 solved?

**Comment by Crimson-foxGITHUB** [comment]
> Was #15 solved?

maybe?

**Comment by Louka3000** [comment]
Great, can you still confirm that no unreal engine games work now? I’ll add a note about it if yes.

**Comment by Crimson-foxGITHUB** [comment]
Tested both Propagation VR and Project Wingman with the upscaler. Both games use the deferred renderer with TAA as an anti-aliasing option, both games had "resolution scale" as options in their graphics menus, and both games would make OpenVR Dynamic Resolution change the global resolution, but absolutely nothing happened in-game.

Deferred rendering in VR is much more common with UE4 games, but Skyrim VR also uses it. Beat Saber in a recent update switched from regular Unity forward rendering to a new mode called Single-Pass Stereo. (a Unity 2021 feature, which might explain why both it and After the Fall don't work)

**Comment by Crimson-foxGITHUB** [comment]
Tested out Bullet Train through Revive. TAA, the app gives resolution changes the game doesn't respond to, same story. Bullet Train doesn't even have a graphics menu, by the way.

**Comment by Crimson-foxGITHUB** [comment]
I tested out Google's WebXR sample using Google Chrome 114 and it was completely unresponsive to global resolution changes. Given that a couple of OpenXR-only titles are listed as "not working" this leads me to believe that any OpenXR-only game will not work with this application.

**Comment by Louka3000** [comment]
I don’t believe OpenXR is a problem, as this can 100% work on post 2019 Unity games (e.g. games using 2021 Unity...)

**Comment by Crimson-foxGITHUB** [comment]
> I don’t believe OpenXR is a problem, as this can 100% work on post 2019 Unity games (e.g. games using 2021 Unity...)

It's a question of which render mode they're using. Worth noting however that OpenVR FSR and VR Performance Toolkit *do* actually work with some UE4 games because they're using a different method to affect resolution.

**Comment by Louka3000** [comment]
Yes, so it has nothing to do with OpenXR.

From what you're saying, it seems deferred rendering is the culprit, but that it doesn't apply to all games from any category, as it's the developer's choice, is that right?

**Comment by Crimson-foxGITHUB** [comment]
> Yes, so it has nothing to do with OpenXR.

> From what you're saying, it seems deferred rendering is the culprit, but that it doesn't apply to all games from any category, as it's the developer's choice, is that right?


Deferred rendering is developer-choice & engine-specific. Deferred rendering has a higher base cost than forward renderer, is much worse for transparency, & doesn't support MSAA, but doesn't need to calculate lighting for every pixel in the entire scene. 

Worth noting that TWD:S&S does use forward rendering but isn't supported...

As for Single-Pass Stereo, I have no explanation as to why it'd cause Beat Saber to break, other than that the game already had a built-in resolution scaler.

**Comment by Crimson-foxGITHUB** [comment]
update: figured out that it's not DX11-only, Vulkan works just fine

**Comment by kruemmelbande** [comment]
I tested population one, and it seemed to not really work, bui I'll do further testing tomorrow
